File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/DotNet/VB.NET/perf_operation/perf_operation.vb
'======================================================================='
'                                                                       '
'                                                                       '
' perf_operation.vb                                                     '
'                                                                       '
'                                                                       '
' Copyright 2008 NetApp, Inc. All rights                                '
' reserved. Specifications subject to change without notice.            '
'                                                                       '
' This SDK sample code is provided AS IS, with no support or            '
' warranties of any kind, including but not limited to                  '
' warranties of merchantability or fitness of any kind,                 '
' expressed or implied.  This code is subject to the license            '
' agreement that accompanies the SDK.                                   '
'                                                                       '
'                                                                       '
'  Sample for usage of following perf group API:                        '
'          perf-object-list-info                                        '
'          perf-object-counter-list-info                                '
'          perf-object-instance-list-info                               '
'          perf-object-get-instances-iter-*                             '
'                                                                       '
' Usage:                                                                '
' perf_operation <filer> <user> <password> <operation>                  '
'                                                                       '
' <filer>      -- Name/IP address of the filer                          '
' <user>       -- User name                                             '
' <password>   -- Password                                              '
' <operation>  --                                                       '
'      object-list - Get the list of perforance objects                 '
'                in the system                                          '
'      instance-list - Get the list of instances for a given            '
'                  performance object                                   '
'      counter-list - Get the list of counters available for a          '
'                 given performance object                              '
'      get-counter-values - get the values of the counters for          '
'                   all instance of a performance object                '
'======================================================================='

Imports System
Imports System.Collections.Generic
Imports NetApp.Manage


Module Perf_Operation
    Public Sub Usage()
        Console.WriteLine(Environment.NewLine & "Usage: perf_operation <filer> <username> <passwd> <operation>")
        Console.WriteLine(Environment.NewLine & "Possible operations are:")
        Console.WriteLine("  object-list - Get the list of perforance objects in the system")
        Console.WriteLine("  instance-list - Get the list of instances for a given performance object")
        Console.WriteLine("  counter-list - Get the list of counters available for a given performance object")
        Console.WriteLine("  get-counter-values - get the values of the counters for all the instances of a performance object")
    End Sub
    Sub Main(ByVal args As String())

        If (args.Length < 4) Then
            Usage()
            Exit Sub
        End If

        Dim serverName As String = args(0)
        Dim user As String = args(1)
        Dim passwd As String = args(2)
        Dim operation As String = args(3)
        Dim objName As String

        Try
            Dim server As New NaServer(serverName, 1, 3)
            server.ServerType = NaServer.SERVER_TYPE.FILER
            server.SetAdminUser(user, passwd)
            Dim input As NaElement
            Dim output As NaElement

            If operation = "object-list" Then
                input = New NaElement("perf-object-list-info")
                output = server.InvokeElem(input)
                Dim objList As IList(Of NaElement) = output.GetChildByName("objects").GetChildren()
                Dim objIter As IEnumerator(Of NaElement) = objList.GetEnumerator()

                While objIter.MoveNext()
                    Dim objInfo As NaElement = objIter.Current
                    Console.Out.Write("Object Name = " + objInfo.GetChildContent("name") + ControlChars.Tab)
                    Console.Out.WriteLine("privilege level = " + objInfo.GetChildContent("privilege-level"))
                    Console.Out.WriteLine()
                End While

            ElseIf operation = "instance-list" Then
                If args.Length < 5 Then
                    Console.Out.WriteLine("Usage:")
                    Console.Out.WriteLine("perf_operation <filer> <user> <password> <instance-list> <objectname>")
                    Environment.Exit(1)
                End If
                input = New NaElement("perf-object-instance-list-info")
                objName = args(4)
                input.AddNewChild("objectname", objName)
                output = server.InvokeElem(input)
                Dim instList As IList(Of NaElement) = output.GetChildByName("instances").GetChildren()
                Dim instIter As IEnumerator(Of NaElement) = instList.GetEnumerator()
                While instIter.MoveNext()
                    Dim instInfo As NaElement = instIter.Current
                    Console.Out.WriteLine("Instance Name = " + instInfo.GetChildContent("name"))
                End While

            ElseIf operation = "counter-list" Then
                If args.Length < 5 Then
                    Console.Out.WriteLine("Usage:")
                    Console.Out.WriteLine("perf_operation <filer> <user> <password> <counter-list> <objectname>")
                    System.Environment.Exit(1)
                End If

                input = New NaElement("perf-object-counter-list-info")
                objName = args(4)
                input.AddNewChild("objectname", objName)
                output = server.InvokeElem(input)
                Dim counterList As IList(Of NaElement) = output.GetChildByName("counters").GetChildren()
                Dim counterIter As IEnumerator(Of NaElement) = counterList.GetEnumerator()

                While counterIter.MoveNext()
                    Dim counterInfo As NaElement = counterIter.Current
                    Console.Out.Write("Counter Name = " + counterInfo.GetChildContent("name"))

                    If counterInfo.GetChildContent("base-counter") <> Nothing Then
                        Console.Out.Write("Base Counter = " + counterInfo.GetChildContent("base-counter"))
                    Else
                        System.Console.Out.Write(", Base Counter = none")
                    End If

                    Console.Out.Write(", Privilege Level = " + counterInfo.GetChildContent("privilege-level"))
                    If counterInfo.GetChildContent("unit") <> Nothing Then
                        Console.Out.Write(", Unit = " + counterInfo.GetChildContent("unit"))
                    Else
                        Console.Out.Write(", Unit = none")
                    End If
                    Console.WriteLine()
                    Console.WriteLine()
                End While

            ElseIf operation = "get-counter-values" Then
                Dim totalRecords As Integer = 0
                Dim maxRecords As Integer = 10
                Dim numRecords As Integer = 0
                Dim iterTag As String = Nothing

                If args.Length < 5 Then

                    System.Console.Out.WriteLine("Usage:")
                    System.Console.Out.WriteLine("perf_operation <filer> <user> <password> <get-counter-values> <objectname> [<counter1> <counter2> ...]")
                    System.Environment.Exit(1)
                End If

                input = New NaElement("perf-object-get-instances-iter-start")
                objName = args(4)
                input.AddNewChild("objectname", objName)
                Dim counters As NaElement = New NaElement("counters")

                'Now store rest of the counter names as child element of counters.
                ' Here it has been hard coded as 5 because first counter is specified at 6th position from 
                ' cmd prompt
                Dim counterIndex As Integer = 5


                While counterIndex < args.Length
                    counters.AddNewChild("counter", args(counterIndex))
                    counterIndex = counterIndex + 1
                End While

                If counterIndex > 5 Then
                    input.AddChildElement(counters)
                End If

                output = server.InvokeElem(input)
                totalRecords = output.GetChildIntValue("records", -1)
                iterTag = output.GetChildContent("tag")

                Do
                    input = New NaElement("perf-object-get-instances-iter-next")
                    input.AddNewChild("tag", iterTag)
                    input.AddNewChild("maximum", System.Convert.ToString(maxRecords))
                    output = server.InvokeElem(input)
                    numRecords = output.GetChildIntValue("records", 0)

                    If numRecords <> 0 Then

                        Dim instList As IList(Of NaElement) = output.GetChildByName("instances").GetChildren()
                        Dim instIter As IEnumerator(Of NaElement) = instList.GetEnumerator()

                        While instIter.MoveNext()

                            Dim instData As NaElement = instIter.Current
                            Console.Out.WriteLine("Instance = " + instData.GetChildContent("name"))

                            Dim counterList As IList(Of NaElement) = instData.GetChildByName("counters").GetChildren()
                            Dim counterIter As IEnumerator(Of NaElement) = counterList.GetEnumerator()

                            While counterIter.MoveNext()
                                Dim counterData As NaElement = counterIter.Current
                                System.Console.Out.Write("counter name = " + counterData.GetChildContent("name"))
                                Console.Out.WriteLine(ControlChars.Tab & "counter value = " + counterData.GetChildContent("value"))
                                Console.Out.WriteLine()
                            End While
                        End While
                    End If
                Loop While numRecords <> 0

                input = New NaElement("perf-object-get-instances-iter-end")
                input.AddNewChild("tag", iterTag)
                output = server.InvokeElem(input)
            Else
                Usage()
            End If
        Catch ex As Exception
            Console.WriteLine(ex.Message)
        End Try
    End Sub

End Module

File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/C/perf_operation/perf_operation.c
//=======================================================================//
//                                                                       //
// $ID$                                                                  //
//                                                                       //
// perf_operation.c                                                      //
//                                                                       //
// Brief information of the contents                                     //
//                                                                       //
// Copyright 2002-2003 Network Appliance, Inc. All rights                //
// reserved. Specifications subject to change without notice.            //
//                                                                       //
// This SDK sample code is provided AS IS, with no support or            //
// warranties of any kind, including but not limited to                  //
// warranties of merchantability or fitness of any kind,                 //
// expressed or implied.  This code is subject to the license            //
// agreement that accompanies the SDK.                                   //
//                                                                       //
//  Sample for usage of following perf group API:                        //
//          perf-object-list-info                                        //
//          perf-object-counter-list-info                                //
//          perf-object-instance-list-info                               //
//          perf-object-get-instances-iter-*                             //
//                                                                       //
// Usage:                                                                //
// perf_operation <filer> <user> <password> <operation>                  //
//                                                                       //
// <filer>      -- Name/IP address of the filer                          //
// <user>       -- User name                                             //
// <password>   -- Password                                              //
// <operation>  --                                                       //
//      object-list - Get the list of perforance objects                 //
//                in the system                                          //
//      instance-list - Get the list of instances for a given            //
//                  performance object                                   //
//      counter-list - Get the list of counters available for a          //
//                 given performance object                              //
//      get-counter-values - get the values of the counters for          //
//                   all instance of a performance object                //
//=======================================================================//


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <netapp_api.h>


void print_usage() {
	printf("\nUsage : perf_operation <filer> <user> <passwd> <operation> [<objectname> <counter1> <counter2> ..]\n");
	printf("<filer>	-- ");
	printf("Name/IP address of the filer\n");
	printf("<user>	-- User Name\n");
	printf("<passwd>	-- Password\n");
	printf("<operation>	--\n");
	printf("  object-list - Get the list of perforance objects in the system\n");
	printf("  instance-list - Get the list of instances for a given performance object\n");
	printf("  counter-list - Get the list of counters available for a given performance object\n");
	printf("  get-counter-values - Get the values of the counters for all the instances of a performance object\n");
	exit(-1);
}


/****************************************************************************
 Name: Main

 Description: Initializes the server context object and calls the
	      appropriate perf API commands based on the input arguments

 Parameters:
      IN:
	argc 	- number of command line arguments
	argv	- Array of command line arguments

 Return value: None
******************************************************************************/

int main(int argc, char* argv[])
{
	na_server_t*    s;
	na_elem_t       *out,*in;
	na_elem_t*	nextElem = NULL;
	na_elem_t*	iterList = NULL;
	na_elem_t*	counterList = NULL;
	na_elem_iter_t	iter, counterIter;
	char*           filerip    = argv[1];
	char*           user   	   = argv[2];
	char*           passwd 	   = argv[3];
	char*           operation  = argv[4];
	char            err[256];

	if (argc < 5) {
		print_usage();
	}

	// One-time initialization of system on client
	if (!na_startup(err, sizeof(err))) {
		fprintf(stderr, "Error in na_startup: %s\n", err);
		return -2;
	}

	//
	// Initialize connection to server, and request version 1.0 of the API set.
	//
	s = na_server_open(filerip, 1, 0);

	//
	// Set connection style (HTTP)
	//
	na_server_style(s, NA_STYLE_LOGIN_PASSWORD);
	na_server_set_transport_type(s,NA_SERVER_TRANSPORT_HTTP, NULL);
	na_server_adminuser(s,user,passwd);

	if(!strcmp(operation, "object-list")) {
		in = na_elem_new("perf-object-list-info");
		out = na_server_invoke_elem(s,in);
		if (na_results_status(out) != NA_OK) {
			printf("Error %d: %s\n", na_results_errno(out),
			na_results_reason(out));
			na_elem_free(in);
			na_elem_free(out);
			na_server_close(s);
			na_shutdown();
			return -3;
		}
		iterList = na_elem_child(out, "objects");
		if(iterList == NULL) {
			// no data to list
			na_elem_free(in);
			na_elem_free(out);
			na_server_close(s);
			na_shutdown();
			return -3;
		}
		printf("\n-------------------------------------------------------\n");
		printf(" Object Name                           Privilege Level");
		printf("\n-------------------------------------------------------\n");
		for (iter=na_child_iterator(iterList); (nextElem =na_iterator_next(&iter)) != NULL;  ) {
			printf(" %-35s    %s\n",
			na_child_get_string(nextElem, "name"),
			na_child_get_string(nextElem, "privilege-level"));
		}
		printf("\n--------------------------------------------------------\n");

	}
	else if(!strcmp(operation, "instance-list")) {

		if( argc < 6) {
			printf("Usage:\n");
			printf("perf_operation <filer> <user> <password> <instance-list> <objectname>\n");
			return -1;
		}

		in = na_elem_new("perf-object-instance-list-info");
		na_child_add_string(in,"objectname",argv[5]);
		out = na_server_invoke_elem(s,in);
		if (na_results_status(out) != NA_OK) {
        	printf("Error %d: %s\n", na_results_errno(out),
			na_results_reason(out));
			na_elem_free(in);
			na_elem_free(out);
			na_server_close(s);
			na_shutdown();
			return -3;
		}
		iterList = na_elem_child(out, "instances");
		if(iterList == NULL) {
			// no data to list
			na_elem_free(in);
			na_elem_free(out);
			na_server_close(s);
			na_shutdown();
			return -3;
		}
		for (iter=na_child_iterator(iterList); (nextElem =na_iterator_next(&iter)) != NULL;  ) {
			printf("Instance Name = %s\n", na_child_get_string(nextElem, "name"));
		}

	}
	else if(!strcmp(operation, "counter-list")) {

		if( argc < 6) {
			printf("Usage:\n");
			printf("perf_operation <filer> <user> <password> <counter-list> <objectname>\n");
			return -1;
		}

		in = na_elem_new("perf-object-counter-list-info");
		na_child_add_string(in,"objectname",argv[5]);

		out = na_server_invoke_elem(s,in);
		if (na_results_status(out) != NA_OK) {
        	printf("Error %d: %s\n", na_results_errno(out),
			na_results_reason(out));
			na_elem_free(in);
			na_elem_free(out);
			na_server_close(s);
			na_shutdown();
			return -3;
		}
		iterList = na_elem_child(out, "counters");
		if(iterList == NULL) {
			// no data to list
			na_elem_free(in);
			na_elem_free(out);
			na_server_close(s);
			na_shutdown();
			return -3;
		}
		for (iter=na_child_iterator(iterList); (nextElem =na_iterator_next(&iter)) != NULL;  ) {

			printf("counter name: %s\n",na_child_get_string(nextElem, "name"));

			if(na_child_get_string(nextElem,"base-counter") != NULL) {
				printf("Base counter: %s\n",na_child_get_string(nextElem,"base-counter"));
			}
			else {
				printf("base counter: none\n");
			}
			printf("privilege-level: %s\n",na_child_get_string(nextElem,"privilege-level"));

			if(na_child_get_string(nextElem,"unit") != NULL) {
				printf("unit: %s\n",na_child_get_string(nextElem,"unit"));
			}
			else {
				printf("Unit: none\n");
			}
			printf("\n");
		}

	}
	else if(!strcmp(operation, "get-counter-values")) {

		int total_records = 0;
		int max_records = 10;
		int num_records = 0;
		char* iter_tag = NULL;
		na_elem_t *counters = NULL;
		int num_counter = 6;

		/* Here num_counter has been hard coded as 6 because first counter
			is specified at 7th position from cmd prompt */

		if( argc < 6) {
			printf("Usage:\n");
			printf("perf_operation <filer> <user> <password> ");
			printf("<get-counter-values> <objectname> [<counter1> <counter2> ..]\n");
			return -1;
		}

		in = na_elem_new("perf-object-get-instances-iter-start");
		na_child_add_string(in,"objectname",argv[5]);

		counters = na_elem_new("counters");

		/*Now store rest of the counter names as child element of counters */
		while(num_counter < argc) {
			na_child_add_string(counters, "counter",argv[num_counter]);
			num_counter++;
		}

		/* If no counters are specified then all the counters are fetched */
		if(num_counter > 6) {
			na_child_add(in,counters);
		}

		out = na_server_invoke_elem(s,in);
		if (na_results_status(out) != NA_OK) {
        	printf("Error %d: %s\n", na_results_errno(out),
			na_results_reason(out));
			na_elem_free(in);
			na_elem_free(out);
			na_server_close(s);
			na_shutdown();
			return -3;
		}
		total_records = na_child_get_int(out,"records",-1);
		iter_tag = (char *)na_child_get_string(out,"tag");

		do {
			in = na_elem_new("perf-object-get-instances-iter-next");
			na_child_add_string(in,"tag",iter_tag);
			na_child_add_int(in,"maximum", max_records);

			out = na_server_invoke_elem(s,in);
			if (na_results_status(out) != NA_OK) {
        		printf("Error %d: %s\n", na_results_errno(out),
				na_results_reason(out));
				na_elem_free(in);
				na_elem_free(out);
				na_server_close(s);
				na_shutdown();
        		return -3;
			}
			num_records = na_child_get_int(out,"records",0);

			if(num_records != 0) {
				iterList = na_elem_child(out, "instances");
				if(iterList == NULL) {
					// no data to list
					na_elem_free(in);
					na_elem_free(out);
					na_server_close(s);
					na_shutdown();
					return -3;
				}
				for (iter=na_child_iterator(iterList); (nextElem =na_iterator_next(&iter)) != NULL;  ) {
					printf("Instance = %s\n", na_child_get_string(nextElem,"name"));
					counterList = na_elem_child(nextElem,"counters");
					if(counterList == NULL) {
						// no data to list
						na_elem_free(in);
						na_elem_free(out);
						na_server_close(s);
						na_shutdown();
						return -3;
					}
					for (counterIter=na_child_iterator(counterList); (nextElem =na_iterator_next(&counterIter)) != NULL;  ) {
						printf("\ncounter name: %s\n",na_child_get_string(nextElem,"name"));
						printf("counter value: %s\n", na_child_get_string(nextElem,"value"));
					}
					printf("\n");
				}
			}
		} while (num_records != 0);

		in = na_elem_new("perf-object-get-instances-iter-end");
		na_child_add_string(in,"tag",iter_tag);
		out = na_server_invoke_elem(s,in);
		if (na_results_status(out) != NA_OK) {
			printf("Error %d: %s\n", na_results_errno(out), na_results_reason(out));
			na_elem_free(in);
			na_elem_free(out);
			na_server_close(s);
			na_shutdown();
			return -3;
		}
	}
	else {
		printf("Invalid Operation \n");
		print_usage();
	}
	na_elem_free(in);
	na_elem_free(out);
	na_server_close(s);
	na_shutdown();
	return 0;
}
